### PR TITLE
[Hackathon] Google Dorking

### PR DESCRIPTION
**Pull request by DeanArbel**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/40600

## Description
Automate the process of google Dorking searches in order to detect leaked data.

### CODE REVIEW

File: Packs/GoogleDorking/Integrations/GoogleDorking/GoogleDorking.py [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+    if max_results % MAX_RESULTS_PER_PAGE:
+        max_pages += 1
+    return max_pages
+
+
+def calculate_page_size(current_page: int, max_results: int) -> Union[int, str]:
+    results_so_far = current_page * MAX_RESULTS_PER_PAGE
+    if (results_diff := max_results - results_so_far) < MAX_RESULTS_PER_PAGE:
+        return results_diff
+    return MAX_RESULTS_PER_PAGE
```
Comment (MosheEichler) [inline Packs/GoogleDorking/Integrations/GoogleDorking/GoogleDorking.py:109]:
```suggestion
    return min(MAX_RESULTS_PER_PAGE, max_results - current_page * MAX_RESULTS_PER_PAGE)
```

File: Packs/GoogleDorking/Scripts/ExtractUsernames/ExtractUsernames.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        raise DemistoException(f"Entry {entry_id} was not found")
+    regex = re.compile(pattern) if pattern else None
+    file_path = res['path']
+    with open(file_path, mode='r') as file:
+        for line in file.readlines():
+            if regex:
+                regex_res = regex.search(line)
+                if regex_res and (user := regex_res.groups()[1]):
+                    users.append(user.lstrip())
+            else:
```
[the remark this answers is not in the document]
Comment (DeanArbel) [reply Packs/GoogleDorking/Scripts/ExtractUsernames/ExtractUsernames.py:22]: No, returning the full line should be ok

File: Packs/GoogleDorking/Scripts/ExtractUsernames/ExtractUsernames.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    users = []
+    res = demisto.getFilePath(entry_id)
+    if not res:
+        raise DemistoException(f"Entry {entry_id} was not found")
+    regex = re.compile(pattern) if pattern else None
+    file_path = res['path']
+    with open(file_path, mode='r') as file:
+        for line in file.readlines():
+            if regex:
+                regex_res = regex.search(line)
```
[the remark this answers is not in the document]
Comment (DeanArbel) [reply Packs/GoogleDorking/Scripts/ExtractUsernames/ExtractUsernames.py:19]: It's simpler to debug this way, we prefer to keep it as is.

File: Packs/GoogleDorking/Integrations/GoogleDorking/GoogleDorking.py [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+        client.urls = argToList(urls)
+    return get_search_results(client, item_to_result)
+
+
+def fetch_incidents(client: Client, last_run: Optional[dict]) -> list:
+    now = datetime.now()
+    last_run_date = last_run.get(LAST_RUN_TIME_KEY) if last_run else None
+    if last_run_date:
+        if (now - dateparser.parse(last_run_date)).days > 0:
+            client.after = last_run_date
```
[the remark this answers is not in the document]
Comment (DeanArbel) [reply Packs/GoogleDorking/Integrations/GoogleDorking/GoogleDorking.py:204]: It's set in the `main` - `client.after = first_fetch_time.strftime(GOOGLE_TIME_FORMAT)`

File: Packs/GoogleDorking/Integrations/GoogleDorking/GoogleDorking.py [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+    def file_types(self, value):
+        self._search_file_types = value
+
+    @property
+    def keywords(self):
+        return self._keywords
+
+    @keywords.setter
+    def keywords(self, value):
+        self._keywords = value
```
[the remark this answers is not in the document]
Comment (DeanArbel) [reply Packs/GoogleDorking/Integrations/GoogleDorking/GoogleDorking.py:72]: https://stackoverflow.com/questions/1568091/why-use-getters-and-setters-accessors

File: Packs/GoogleDorking/Integrations/GoogleDorking/GoogleDorking.py [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+    if 'link' in item:
+        link = item['link']
+        file_result = fileResult(path.basename(link), client._http_request('GET', full_url=link, resp_type='content'))
+        results.append(file_result)
+    results.append(
+        CommandResults(
+            outputs_prefix='GoogleDorking',
+            outputs=item
+        )
+    )
```
[the remark this answers is not in the document]
Comment (DeanArbel) [reply Packs/GoogleDorking/Integrations/GoogleDorking/GoogleDorking.py:172]: yeah, the `return_results` function iterates over the results and calls `demisto.results` for each one - so it's possible to have different item types in the same list.

File: Packs/GoogleDorking/Integrations/GoogleDorking/GoogleDorking.py [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+
+        for item in items:
+            results.extend(parser(client, item))
+
+        # prepare next run
+        current_page += 1
+        start += results_in_page
+        total = int(demisto.get(page, 'searchInformation.totalResults',
+                                demisto.get(page, 'queries.request.totalResults', 0)))
+        all_pages_found = total < start
```
[the remark this answers is not in the document]
Comment (DeanArbel) [reply Packs/GoogleDorking/Integrations/GoogleDorking/GoogleDorking.py:135]: It does reach that point, but it's generally a good practice to check `le`. I'll adjust it.

**Review by MosheEichler — APPROVED** [review]
Perfect!

**Pull request merged**